Feature/Jest to Vitest migration (#6605)
plone/volto: docs/source/development/add-ons/test-add-ons-18.md

This guide assumes that you've used {term}`Cookieplone` to create your add-on boilerplate.
```

This chapter describes how to migrate your tests from Jest to [Vitest](https://vitest.dev/guide/).
Although Volto core now uses Vitest as its unit test runner, projects generated through {term}`Cookieplone` use Jest.
Cookieplone will have a support for Vitest in the near future.

```{versionadded} Volto 18.TBD.TBD, current release at 18.10.1
Volto core now uses Vitest for its unit tests.
```

```{deprecated} Volto 18.TBD.TBD, current release at 18.10.1
Volto core has migrated from Jest to Vitest for unit tests.
Jest is now deprecated.
```

## Install Vitest

Run the following command to add Vitest and required dependencies to your development environment.

```shell
pnpm add -D vitest @testing-library/react jsdom
```

> ## Configure your add-on to use Vitest

Reviewer: This is gold! thanks!

```

In Vitest, `require()` is not supported, and `import()` may sometimes fail.
In such cases, `jiti` provides a reliable alternative for module resolution and execution.
In Volto, `jiti` is used in the test file {file}`packages/volto/__tests__/create-addons-loader.test.js` to handle dynamic imports for add-ons efficiently.

## Additional Vitest information

```{seealso}
For a general guide for Vitest, see {doc}`../../contributing/testing`.
```
For complete details on migrating from Jest to Vitest, refer to the official [Vitest Migration Guide](https://vitest.dev/guide/migration.html#jest).

## Jest for Volto add-ons

Volto previously used {term}`Jest` as its unit testing framework.
If your add-on still uses Jest, Volto provides backward compatibility.

While Volto core has migrated to Vitest, it retained its Jest configuration, allowing you to continue running your existing Jest tests in your add-on without issues.

This support lets you migrate your tests from Jest to Vitest at your convenience.

### Jest configuration override

If you use Jest for your add-on testing, you may need to customize the configuration, especially in GitHub workflows or local development.

> Volto provides a way to do this by either using a {file}`jest.config.js` file, or specifying a custom configuration file through the `RAZZLE_JEST_CONFIG` environment variable.

Reviewer 2: I recently found the [`envvar` role in Sphinx](https://www.sphinx-doc.org/en/master/usage/referencing.html#role-envvar).

```suggestion
Volto provides a way to do this by either using a {file}`jest.config.js` file, or specifying a custom configuration file through the {envvar}`RAZZLE_JEST_CONFIG` environment variable.
```

```shell
RAZZLE_JEST_CONFIG=my-custom-jest-config.js pnpm test
```

```{note}
Both configurations are merged in a way that the keys of the configuration provided override the initial {file}`package.json` default configuration, either in Volto or in your projects.
```

This is especially useful in GitHub workflows while developing add-ons.
You can pass a specific configuration file that properly deals with the add-on configuration.

> ### Add add-ons via environment variable for testing purposes

Reviewer 2: ```suggestion
### Add add-ons via environment variable
```

> Sometimes you need to enable different configurations and enable optional components, for example, testing purposes.

Reviewer 2: ```suggestion
Sometimes you need to enable different configurations and enable optional components, such as for testing purposes.
```

> You can use the `ADDONS` environment variable to define them.

Reviewer 2: ```suggestion
You can use the {envvar}`ADDONS` environment variable to define them.
```

```bash
ADDONS=test-addon,test-addon2 pnpm start
```

See {doc}`../configuration/environmentvariables` for more information.

### Override Jest configuration

In {term}`CI` or for testing add-ons, it's useful to modify Jest's {file}`package.json` configuration file.
You can use the file {file}`jest.config.js` provided by the boilerplate.
The test command will load it and apply it.

```{warning}
Do not modify the existing keys in there if you don't know what you're doing.
Some of them are required for the tests to run properly in the Volto context.
```

> Both configurations are merged in a way that the keys of the configuration provided override the initial {file}`package.json` configuration, either in Volto or in your projects.

Reviewer 2: ```suggestion
Both configurations are merged in such a way that the keys of the provided configuration override the initial {file}`package.json` configuration, in either Volto or your projects.
```

> 
> 

Reviewer 2: ```suggestion
```

This guide assumes that you've used {term}`Cookieplone` to create your add-on boilerplate.
```

> This chapter describes how to set up your testing environment and to write tests using [Vitest](https://vitest.dev/guide/), and how to migrate your tests from Jest to Vitest.

Reviewer 2: ```suggestion
This chapter describes how to set up your testing environment and to write tests using [Vitest](https://vitest.dev/guide/) for your add-on, and how to migrate your add-on's tests from Jest to Vitest.
```

  jest.fn(() => <div id="state-select" />),
);
```

The Vitest equivalent is the following example.

```javascript
vi.mock('../../manage/Workflow/Workflow', () => ({
  default: vi.fn(() => <div id="state-select" />),
}));
```

Vitest's `vi.mock()` does not automatically assume a default export like Jest does.
Instead, it treats the module as an object, so the mocked function must be explicitly assigned to the `default` property.
For more details, refer to the [Mocking](https://vitest.dev/guide/mocking.html) guide in the Vitest documentation.

### Testing with lazy loaded libraries

In Jest, you would test lazy loaded libraries as shown.

```javascript
jest.mock('@plone/volto/helpers/Loadable/Loadable');
beforeAll(
  async () =>
    await require('@plone/volto/helpers/Loadable/Loadable').__setLoadables(),
);
```

In Vitest the equivalent is the following example.

```javascript
vi.mock('@plone/volto/helpers/Loadable/Loadable');
beforeAll(async () => {
  const { __setLoadables } = await import('@plone/volto/helpers/Loadable/Loadable');
  await __setLoadables();
});
```

`require()` is not supported in Vitest.
Instead, use `import()`.

> 

Reviewer 2: ```suggestion

### Additional Jest to Vitest migration information

For complete details on migrating from Jest to Vitest, refer to the official [Vitest Migration Guide](https://vitest.dev/guide/migration.html#jest).

```

## Run tests in Vitest

After completing the foregoing steps, you can use Vitest.
To run your tests, execute the following command.

```shell
make test
```

## `jiti` for add-on testing

[`jiti`](https://www.npmjs.com/package/jiti) is a module loader for JavaScript and TypeScript that allows you to import and execute TypeScript files or ESM modules without requiring a build step.
It is particularly useful in Volto for dynamically loading add-ons and their configurations.

To use `jiti`, see the following code example.

```javascript
import jiti from 'jiti';

const _import = jiti(import.meta.url, { esmResolve: true });
```

In Vitest, `require()` is not supported, and `import()` may sometimes fail.
In such cases, `jiti` provides a reliable alternative for module resolution and execution.
In Volto, `jiti` is used in the test file {file}`packages/volto/__tests__/create-addons-loader.test.js` to handle dynamic imports for add-ons efficiently.

## Additional Vitest information

> For complete details on migrating from Jest to Vitest, refer to the official [Vitest Migration Guide](https://vitest.dev/guide/migration.html#jest).
> 

Reviewer 2: Moved to better context.

```suggestion
```

This guide assumes that you've used {term}`Cookieplone` to create your add-on boilerplate.
```

> This chapter describes how to set up your testing environment and to write tests using [Vitest](https://vitest.dev/guide/) for your add-on, and how to migrate your add-on's tests from Jest to Vitest.

Reviewer 2: ```suggestion
This chapter describes how to set up your testing environment and to write tests using [Vitest](https://vitest.dev/guide/) for your add-on, and how to migrate your add-on's tests from Jest to Vitest, in Volto 18.
```

The second alias resolves the `promise-file-reader` dependency.

```{tip}
If your add-on previously relied on Jest's configuration via the `RAZZLE_JEST_CONFIG` environment variable, you can now remove it, as Vitest does not use it.
```

## Create setup file for Vitest

Create a test setup file {file}`setupTests.js` for Vitest.
The following code is boilerplate setup for Vitest.

```javascript
import '@testing-library/jest-dom';
import { expect, describe, it, vi } from 'vitest';

// Make Vitest globals available throughout the test suite
global.describe = describe;
global.it = it;
global.expect = expect;
global.vi = vi;

// Stub the global fetch API to prevent actual network requests in tests
vi.stubGlobal('fetch', vi.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
    text: () => Promise.resolve(''),
  })
));

// Add additional configurations as needed
```

## Update {file}`package.json` to use Vitest

To switch your project from Jest to Vitest, update the file {file}`package.json`.

```json
"scripts": {
  "test": "vitest",
  "test:watch": "vitest --watch",
  "test:coverage": "vitest --coverage"
}
```

> ## Migrate from Jest to Vitest

Author: @stevepiercy I think we should move this title on top of `install vitest`as this doc is more about migrating to vitest and adjust section below title accordingly. What do u think?

Reviewer 2: Yeah, and we'll need to restructure this again after Cookieplone gets Vitest support. Let me edit this until it flows better. I'll just move sections around and change heading levels.

Author: @stevepiercy You can replace this title by "Similarity between Jest and Vitest " or something like that and move that Migrate from Jest to Vitest Title to top

Reviewer 2: @Abhishek-17h @sneridagh please take a look at this commit https://github.com/plone/volto/pull/6605/commits/e5194a3c7db12d36a14ae088ce9f5177d4fcd34d. I think this improves the flow of the documentation, but I've been staring at this for over an hour, and I'm too tired to see this clearly. Another pair of eyes and a clear head would be helpful. Thank you!

Reviewer 2: It would probably make more sense to wait for the RTD build to complete and preview it at https://volto--6605.org.readthedocs.build/6605/development/add-ons/test-add-ons-18.html
